solr/core/src/java/org/apache/solr/search/ExtendedDismaxQParser.java

```diff
@@ -199,6 +199,18 @@ public Query parse() throws SyntaxError {
       topQuery = FunctionScoreQuery.boostByValue(topQuery, boosts.get(0).asDoubleValuesSource());
     }
 
+    // If topQuery is a boolean query, unwrap the boolean query to check if it is just
+    // a MatchAllDocsQuery. Using MatchAllDocsQuery by itself enables later optimizations
+    if (topQuery instanceof BooleanQuery) {
+      BooleanQuery topQueryBoolean = (BooleanQuery) topQuery;
+      if (topQueryBoolean.clauses().size() == 1) {
+        Query onlyQuery = topQueryBoolean.clauses().get(0).getQuery();
+        if (onlyQuery instanceof MatchAllDocsQuery) {
```

Comment:
There are probably other cases besides `MatchAllDocsQuery` that could be unwrapped here. Another example being `ConstantScoreQuery`.